### PR TITLE
decouple image name fetching from image puller

### PR DESCRIPTION
**Pull request by leoyifeng**
1. decouple image name fetching from image puller to allow customization of fetching image names process
2. change docker build to be multi arch compatible using apt-get install instead of curl amd64 tar

### CODE REVIEW

**Review by kevin-bates — COMMENTED** [review]
Hi @leoyifeng - thank you for the PR - this makes good sense.

Since you're adding additional KIP environment/configuration variables, I think it would be good to have these documented in the [helm chart files](https://github.com/jupyter-server/enterprise_gateway/blob/aa2e015d207fe31bac5a1b330ef3381a11de2cee/etc/kubernetes/helm/enterprise-gateway/templates/daemonset.yaml#L39) and in the docs in the [table describing the helm chart variables](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/operators/deploy-kubernetes.html) and [where we discuss KIP deployment](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/operators/deploy-kubernetes.html#helm-release-to-deploy-kernel-image-puller).

File: etc/docker/kernel-image-puller/image_fetcher.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        fetchers (list): A list of fetchers to use for fetching image names.
+
+    Methods:
+        fetch_images(): Fetches image names from all fetchers and returns them as a set.
+    """
+
+    def __init__(self, logger):
+        self.logger = logger
+        fetcher_names = os.getenv("KIP_INTERNAL_FETCHERS", "KernelSpecsFetcher").split(',')
+        self.fechers = []
```
[the remark this answers is not in the document]
Comment (leoyifeng) [reply etc/docker/kernel-image-puller/image_fetcher.py]: good catch @kevin-bates, will update the documentation

**Comment by leoyifeng** [comment]
@kevin-bates could you advise why the build failed ? thx

**Comment by kevin-bates** [comment]
Hi @leoyifeng - I just pushed a couple of updates to address the lint failures.  However, after this, there's a lint failure on the `pipx run interrogate -v` step of `make lint` - which is failing.  I'm not that familiar with that particular command, but it appears to look for documentation deficiencies.  Not sure why it is complaining about the new file though.

**Comment by kevin-bates** [comment]
Hi @leoyifeng,

It looks like `interrogate` is complaining about missing docstrings on the `fetch_image_names` and the `__init__` methods.

If you `pip install interrogate` then run `interrogate -vv .` (from the root of the repo), it will show the methods that are not "adequately" doc-stringed.

**Comment by leoyifeng** [comment]
@kevin-bates thx for the pointer, now all checks passed. appreciate it.

**Comment by kevin-bates** [comment]
Hi @leoyifeng - this is looking good.  I think once the new `kip` variables are added to [this table](https://github.com/jupyter-server/enterprise_gateway/blob/main/docs/source/operators/deploy-kubernetes.md#helm-configuration-parameters) we should be good to go.  Thanks.

**Review by lresende — APPROVED** [review]
Thanks, @leoyifeng for these nice improvements which will provide a lot more flexibility for the gateway image puller. LGTM after we complete the doc updates suggested by @kevin-bates above.

**Review by kevin-bates — APPROVED** [review]
Thank for your the great contribution @leoyifeng!

**Pull request merged**